### PR TITLE
T28 messages

### LINKED ISSUE #28: Refactor Phalcon\Messages
Refactor Phalcon\Messages

### PR DESCRIPTION
**Pull request by tacxou**
Hello!

* Type: new feature
* Link to issue: #28 

**In raising this pull request, I confirm the following:**

- [x] I have read and understood the [Contributing Guidelines](https://github.com/phalcon/phalcon/blob/master/CONTRIBUTING.md)
- [x] I have checked that another pull request for this purpose does not exist
- [x] I wrote some tests for this PR
- [ ] I have updated the relevant CHANGELOG
- [ ] I have created a PR for the [documentation](https://github.com/phalcon/docs) about this change

Thanks

### CODE REVIEW

**Comment by tacxou** [comment]
Need to separate access arrays in traits. Working...

File: src/Messages/Messages.php [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+     * $messages[0] = new \Phalcon\Messages\Message("This is a message");
+     *```
+     *
+     * @param mixed $offset
+     * @param Message $message
+     * @throws Exception
+     */
+    public function offsetSet($offset, $message): void
+    {
+        if (!is_object($message)) {
```
Comment (niden) [inline src/Messages/Messages.php:232]: Should this be an `is_object` check or a check for type `MessageInterface`?

File: src/Messages/Messages.php [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+             */
+
+            $messages->rewind();
+
+            while ($messages->valid()) {
+                $message = $messages->current();
+                $this->appendMessage($message);
+                $messages->next();
+            }
+        }
```
Comment (niden) [inline src/Messages/Messages.php:98]: I think we need to change this a bit. If the array does not have Message objects then we will store different stuff in the collection. Also, I do not remember why I did the `else` branch that way, most likely to update the `position` property but we can simplify this code - I doubt that this is optimized. 

Thoughts?

**Review by niden — CHANGES_REQUESTED** [review]

File: src/Messages/Messages.php (hunk 1 above)
Comment (tacxou) [reply src/Messages/Messages.php:232]: MessageInterface, I didn't dare to change the existing code

File: src/Messages/Messages.php (hunk 2 above)
Comment (tacxou) [reply src/Messages/Messages.php:98]: If the object is iterable, we can foreach and unify the code between the table and the object

Comment (tacxou) [reply src/Messages/Messages.php:98]: One could also check the MessageInterface at iteration

**Comment by niden** [comment]
@tacxticx88 I think what we should do is leave this as is, exact translation from zephir to php, but create a couple of issues that reference the above conversation (iterable, messageinterface etc.).

Then we can address those in a future version.

**Comment by tacxou** [comment]
Ok, I take note, can I validate like that then?

**Comment by niden** [comment]
> Ok, I take note, can I validate like that then?

Yes this should be OK as is (Zephir -> PHP). We can make it better in the next version

**Review by niden — APPROVED** [review]

**Pull request merged**